wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/environment/InternalEnvironmentFactory.java

```diff
@@ -0,0 +1,135 @@
+ * </pre>
+ *
+ * @author gazarenkov
+ * @author Sergii Leshchenko
+ */
+public abstract class InternalEnvironmentFactory<T extends InternalEnvironment> {
+
+  private final InstallerRegistry installerRegistry;
+  private final RecipeRetriever recipeRetriever;
+  private final InternalEnvironmentValidator<T> envValidator;
+
+  public InternalEnvironmentFactory(
+      InstallerRegistry installerRegistry,
+      RecipeRetriever recipeRetriever,
+      InternalEnvironmentValidator<T> envValidator) {
+    this.installerRegistry = installerRegistry;
+    this.recipeRetriever = recipeRetriever;
+    this.envValidator = envValidator;
+  }
+
+  /**
+   * Creates a valid instance of InternalEnvironment.
+   *
+   * @param sourceEnv the environment
+   * @return InternalEnvironment a valid InternalEnvironment instance
+   * @throws InfrastructureException if exception occurs on recipe downloading
+   * @throws InfrastructureException if infrastructure specific error occurs
+   * @throws ValidationException if validation fails
+   */
+  public T create(final Environment sourceEnv) throws InfrastructureException, ValidationException {
+
+    Map<String, InternalMachineConfig> machines = new HashMap<>();
+    List<Warning> warnings = new ArrayList<>();
+    if (sourceEnv.getWarnings() != null) {
+      warnings.addAll(sourceEnv.getWarnings());
+    }
+
+    InternalRecipe recipe = recipeRetriever.getRecipe(sourceEnv.getRecipe());
+
+    for (Map.Entry<String, ? extends MachineConfig> machineEntry :
+        sourceEnv.getMachines().entrySet()) {
+      MachineConfig machineConfig = machineEntry.getValue();
+
+      List<Installer> installers;
+      try {
+        installers = installerRegistry.getOrderedInstallers(machineConfig.getInstallers());
+      } catch (InstallerException e) {
+        throw new InfrastructureException(e);
+      }
+
+      machines.put(
+          machineEntry.getKey(),
+          new InternalMachineConfig(
+              installers,
+              normalizeServers(machineConfig.getServers()),
+              machineConfig.getEnv(),
+              machineConfig.getAttributes()));
+    }
+
+    T env = doCreate(recipe, machines, warnings);
+
+    envValidator.validate(env);
+
+    return env;
+  }
+
+  /**
+   * Implementation validates downloaded recipe and creates specific InternalEnvironment.
+   *
+   * @param recipe downloaded recipe
+   * @param machines machines configuration
+   * @param warnings list of warnings
+   * @return InternalEnvironment
+   * @throws InfrastructureException if infrastructure specific error occurs
+   * @throws ValidationException if validation fails
+   */
+  protected abstract T doCreate(
+      InternalRecipe recipe, Map<String, InternalMachineConfig> machines, List<Warning> warnings)
+      throws InfrastructureException, ValidationException;
+
+  @VisibleForTesting
+  Map<String, ServerConfig> normalizeServers(Map<String, ? extends ServerConfig> servers) {
```

Comment:
This method might be useful for concrete impls of this class, so I would make it protected.

Reply:
Or method `normalizeServer`

Reply:
We removed functionality when servers can be added depending on the recipe. So I don't think that it should be protected for now. Let's make it protected when if it will be needed.